qiskit/synthesis/evolution/product_formula.py

```diff
@@ -177,6 +186,79 @@ def evolve_pauli(
         _multi_qubit_evolution(output, pauli, time, cx_structure, wrap)
 
 
+def reorder_paulis(
+    operators: SparsePauliOp | list[SparsePauliOp],
+) -> SparsePauliOp | list[SparsePauliOp]:
+    r"""
+    Creates an equivalent operator by reordering terms in order to yield a
+    shallower circuit after evolution synthesis. The original operator remains
+    unchanged.
+
+    This method works in three steps. First, a graph is constructed, where the
+    nodes are the terms of the operator and where two nodes are connected if
+    their term acts on the same qubit (for example, the terms $IXX$ and $IYI$
+    would be connected, but not $IXX$ and $YII$). Then, the graph is colored.
+    Two terms with the same color thus do not act on the same qubit, and in
+    particular, their evolution subcircuits can be run in parallel in the
+    greater evolution circuit of ``operator``. Finally, a new
+    :class:`~qiskit.quantum_info.SparsePauliOp` is created where terms of the
+    same color are grouped together.
+
+    In trivial cases, i.e. when either
+    - the input is a :class:`~qiskit.quantum_info.SparsePauliOp` with
+      less than two Pauli terms, or
+    - the input is a list containing a single
+      :class:`~qiskit.quantum_info.SparsePauliOp` which has less than two
+      Pauli terms,
+    this method does nothing. In non-trivial cases, this method always returns a
+    :class:`~qiskit.quantum_info.SparsePauliOp` (rather than a list of such),
+    even if ``operators`` is a list.
+
+    This method is deterministic and invariant under permutation of the Pauli term in ``operators``.
+
+    Args:
+        operators: The operator or list of operators whose terms to reorder.
+    """
+
+    def _term_sort_key(term: tuple[str, complex]) -> Any:
+        return (term[0], term[1].real, term[1].imag)
+
+    # Do nothing in trivial cases
+    if isinstance(operators, SparsePauliOp) and len(operators) <= 1:
+        return operators
+    if isinstance(operators, list) and not (
+        # operators is a list of > 1 SparsePauliOp
+        len(operators) > 1
+        # operators has a single SparsePauliOp, which has > 1 terms
+        or (len(operators) == 1 and len(operators[0]) > 1)
+    ):
+        return operators
```

Comment:
Pauli operators that are given in a list all commute, e.g. we would give
```
[SparsePauliOp(["ZZI", "IZZ", "ZII"]), SparsePauliOp(["YII", "YYY"])]
```
This information should be preserved if provided by the users 🙂 So I think we should simply do a re-ordering per element in the list, instead of merging all operators and discarding this initial information. If users would like to merge all operators, this could simply be done by adding the operators before passing them into the evolution.

Reply:
I agree that preserving this information is needed. I suppose you could simply call the algorithm below on each entry of this list separately and then return a new list of operators.